core/store/src/trie/shard_tries.rs

```diff
@@ -13,18 +13,28 @@ use near_primitives::types::{
 
 use crate::db::{DBCol, DBOp, DBTransaction};
 use crate::trie::trie_storage::{TrieCache, TrieCachingStorage};
-use crate::trie::TrieRefcountChange;
+use crate::trie::{TrieRefcountChange, POISONED_LOCK_ERR};
 use crate::{StorageError, Store, StoreUpdate, Trie, TrieChanges, TrieUpdate};
 
-#[derive(Clone)]
-pub struct ShardTries {
-    pub(crate) store: Arc<Store>,
+struct ShardTriesInner {
+    store: Arc<Store>,
     /// Cache reserved for client actor to use
-    pub(crate) caches: Arc<HashMap<ShardUId, TrieCache>>,
+    caches: RwLock<HashMap<ShardUId, TrieCache>>,
     /// Cache for readers.
-    pub(crate) view_caches: Arc<HashMap<ShardUId, TrieCache>>,
+    view_caches: RwLock<HashMap<ShardUId, TrieCache>>,
+}
+
+#[derive(Clone)]
+pub struct ShardTries(Arc<ShardTriesInner>);
+
+impl PartialEq for ShardTries {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.0, &other.0)
+    }
 }
 
+impl Eq for ShardTries {}
+
```

Comment:
This looks a bit odd. Identity semantics is at times very useful in Rust, but it usually indicates that something really important is going on, and that the caller is supposed to actively use the identity by, eg, storing things in a hash map. As there's only one call-site which makes use of this, I suggest starting with special casing instead of `impl Eq`:

```rust
impl ShardTries {
  pub fn is_same(&self, other: &ShardTries) -> bool
}
```